### ISSUE TITLE
the loop variable is used to index collection when it is used variadically

### ISSUE THREAD

**Issue by robbiemu** [rust-lang/rust-clippy#11344]
### Summary

given 
```rust
      for i in 0..4 {
        if let Some(index) = adjacencies[i] {
          let Some(position) = mapping.get_mapping_location_of_index(index) else {
            return cube;
          };
          if !visited.contains(&position) {
            let heading = cube.get_turn(&path).unwrap_or(0);
            let cardinal_index = (i as isize + heading).rem_euclid(4) as usize;

            let result = cube.add_to(
              CubeFace::from_cardinal_index(cardinal_index),
              position.to_owned(),
            );
```

here i is used both as a location in adjacencies, and as a rotational offset from the current position (cardinal_index initialization). when the loop variable is used (in a variable or otherwise) other than to index a collection, it should remove the warning.

### Lint Name

clippy::needless_range_loop

### Reproducer

I tried this code from the summary

I saw this happen:

```
warning: the loop variable `i` is used to index `adjacencies`
   --> src/cube/mod.rs:895:16
    |
895 |       for i in 0..4 {
    |                ^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_range_loop
    = note: `#[warn(clippy::needless_range_loop)]` on by default
help: consider using an iterator and enumerate()
    |
895 |       for (i, <item>) in adjacencies.iter().enumerate() {
    |           ~~~~~~~~~~~    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

I expected to see this happen:
no warning

### Version

```text
rustc 1.71.0 (8ede3aae2 2023-07-12)
binary: rustc
commit-hash: 8ede3aae28fe6e4d52b38157d7bfe0d3bceef225
commit-date: 2023-07-12
host: x86_64-apple-darwin
release: 1.71.0
LLVM version: 16.0.5
```

**Comment by Centri3** [comment]
Why wouldn't `.iter().enumerate()` work? This gives you both `i` and `index` without any indexing or unwrapping. This will only ever be `0..4` since this is an array (I think, iirc it only suggests this on arrays. Otherwise it uses `.iter().take()`). imo it's also cleaner:

```rust
    for (i, index) in adjacencies.iter().enumerate() {
        let Some(position) = mapping.get_mapping_location_of_index(index) else {
            return cube;
        };
        if !visited.contains(&position) {
            let heading = cube.get_turn(&path).unwrap_or(0);
            let cardinal_index = (i as isize + heading).rem_euclid(4) as usize;

            let result = cube.add_to(
                CubeFace::from_cardinal_index(cardinal_index),
                position.to_owned(),
            );
        }
    }
````

This shouldn't cause compile errors since it's an iterator and you're indexing into an array (which it checks for).
